build-logic/basics/src/main/kotlin/gradlebuild.minify.gradle.kts

```diff
@@ -72,19 +73,17 @@ plugins.withId("java-base") {
             }
         }
     }
-    configurations.all {
-        // TODO we should be able to solve this only by requesting attributes for artifacts - https://github.com/gradle/gradle/issues/11831#issuecomment-580686994
-        (this as ConfigurationInternal).beforeLocking {
-            // everywhere where we resolve, prefer the minified version
-            if (isCanBeResolved && !isCanBeConsumed && name !in setOf("currentApiClasspath", "codenarc")) {
-                attributes.attribute(minified, true)
-            }
-            // local projects are already minified
-            if (isCanBeConsumed && !isCanBeResolved) {
-                if (attributes.getAttribute(Category.CATEGORY_ATTRIBUTE)?.name == Category.LIBRARY
-                    && attributes.getAttribute(Bundling.BUNDLING_ATTRIBUTE)?.name == Bundling.EXTERNAL
-                ) {
-                    attributes.attribute(minified, true)
+    afterEvaluate {
+        // Without afterEvaluate, configurations.all runs before the configurations' roles are set.
+        // This is yet another reason we need configuration factory methods.
+        configurations.all {
+            if (isCanBeResolved && !isCanBeConsumed) {
+                resolutionStrategy.dependencySubstitution {
+                    substitute(module("it.unimi.dsi:fastutil")).using(variant(module("it.unimi.dsi:fastutil:8.5.2")) {
```

Comment:
Dependency substitution rules _require_ that you include a version.

Reply:
* See #15181

Reply:
Closed as not planned..... This is clearly a hole in the API, and while it may not currently be on our radar I think its worth having an issue to track it and for users to upvote.

Reply:
Reopened this issue